Updated API stability policy (#19599)
nvaccess/nvda: projectDocs/dev/developerGuide/developerGuide.md

### Add-on API stability {#API}

#### Definition of the public API

The NVDA Add-on API consists of all NVDA Python objects, classes and functions, excluding symbols prefixed with an underscore (`_`), transitive imports and Python packages.

* **Public symbols (e.g. `gui.mainFrame`):** These form the supported contract.
We strive to maintain backward compatibility for these symbols according to the schedule below.
* **Private symbols (e.g. `_doInstall`):** These are internal implementation details.
They may change or disappear in any release without notice.
Add-ons relying on private symbols do so at their own risk.

> This also includes anything "below" the underscore.

Reviewer: I suggest an example here, such as:

```suggestion
This also includes anything "below" the underscore.
    + For example, in the import `package._module.class`, both `_module` and `class`, are private.
```

* **Examples:** Fixing a math error or optimising an algorithm (e.g. changing a sort order).
* **Policy:** These are permitted in any major or minor release (e.g. 2026.1, 2026.2).
* We recognise that some add-ons may inadvertently rely on incorrect behaviour / bugs.
However, preserving a bug to maintain backwards compatibility significantly hampers NVDA's development.
If a public function was documented to do X but was actually doing Y, changing it to correctly do X is considered a bug fix, not a breaking change.
* Conversely, changing a function that was documented to do X, and was correctly doing X, to instead do Y is considered a breaking change.

##### Security improvements

Security takes precedence over backward compatibility.
We will make every effort to document these changes clearly in the changelog.

* **Examples:** Privilege escalation or insecure file handling.
* **Policy:** These are permitted in any release but are primarily deployed in patch releases (e.g. 2026.1.2).

#### API deprecation strategy

We im to avoid silent breaks of add-on code.

> Wh  removing or changing APIs in a breaking release (e.g. 2026.1):

Reviewer 2: ```suggestion
We aim to avoid silent breaks of add-on code.
When removing or changing APIs in a breaking release (e.g. 2026.1):
```

### Add-on API stability {#API}

#### Definition of the public API

The NVDA Add-on API consists of all NVDA Python objects, classes and functions, excluding symbols prefixed with an underscore (`_`), transitive imports and Python packages.

* **Public symbols (e.g. `gui.mainFrame`):**
These form the supported contract.
We strive to maintain backward compatibility for these symbols according to the schedule below.
* **Private symbols (e.g. `_doInstall`):**
These are internal implementation details.
Code is considered private if it begins with a leading underscore.
This also includes anything "below" the underscore.
For example, `package.module._Class`, `package._module.Class`, and `_package.module.Class` are all internal.
They may change or disappear in any release without notice.
Add-ons relying on private symbols do so at their own risk.
* **Pip packages:**
These may be updated, downgraded or removed at any time.
It is recommended to package any pip dependency directly with your add-on, rather than using NVDA's version of the package.

> The exception to this is `wx`, which must be used from NVDA.

Reviewer 2: ```suggestion
The exception to this is `wxPython`, which must be used from NVDA.
```

#### The API release cycle

To balance progress with ecosystem stability, NVDA follows a predictable schedule for API changes:

* **Annual API-breaking release (major releases, e.g. 2026.1):**
This is the only release window where signature breaking changes (defined below) are permitted.
This allows us to remove technical debt and refactor core code.
* **Standard releases (minor & patch releases, e.g. 2026.2, 2026.3.1):**
These releases focus on features and stability.
Existing API signatures are preserved, but new API features may be added. 
This includes behavioural refinements and bug fixes (see below).

#### Classification of API changes

To provide clarity on what add-on developers can expect, we categorise changes into three tiers:

##### Signature breaking changes

* **Examples:** Renaming a function, removing a module, changing a function’s return type or removing/reordering positional arguments.
* **Policy:** These occur only in the .1 release (e.g. 2026.1).

##### Behavioural refinements & bug fixes

> A change where the code continues to run (the signature matches), but the logic or outcome changes to correct a defect or security flaw.

Reviewer 2: ```suggestion
##### Behavioural refinements & bug fixes

A change where the code continues to run (the signature matches), but the logic or outcome changes to correct a defect or security flaw.
```

Code is considered private if it begins with a leading underscore.
This also includes anything "below" the underscore.
For example, `package.module._Class`, `package._module.Class`, and `_package.module.Class` are all internal.
They may change or disappear in any release without notice.
Add-ons relying on private symbols do so at their own risk.
* **Pip packages:**
These may be updated, downgraded or removed at any time.
It is recommended to package any pip dependency directly with your add-on, rather than using NVDA's version of the package.
The exception to this is `wx`, which must be used from NVDA.

#### The API release cycle

To balance progress with ecosystem stability, NVDA follows a predictable schedule for API changes:

* **Annual API-breaking release (major releases, e.g. 2026.1):**
This is the only release window where signature breaking changes (defined below) are permitted.
This allows us to remove technical debt and refactor core code.
* **Standard releases (minor & patch releases, e.g. 2026.2, 2026.3.1):**
These releases focus on features and stability.
Existing API signatures are preserved, but new API features may be added. 

> Behavioural refinements and bug fixes are also permitted (see below).

Author: Accepted with slightly different phrasing.

#### Classification of API changes

To provide clarity on what add-on developers can expect, we categorise changes into three tiers:

##### Signature breaking changes

* **Examples:** Renaming a function, removing a module, changing a function’s return type or removing/reordering positional arguments.
* **Policy:** These occur only in the .1 release (e.g. 2026.1).

##### Behavioural refinements & bug fixes
A change where the code continues to run (the signature matches), but the logic or outcome changes to correct a defect or security flaw.

* **Examples:** Fixing a math error or optimising an algorithm (e.g. changing a sort order).
* **Policy:** These are permitted in any major or minor release (e.g. 2026.1, 2026.2).

> * We recognise that some add-ons may inadvertently rely on incorrect behaviour / bugs.

Reviewer: Reliance on known behaviors, though incorrect, may be entirely intentional.

```suggestion
* We recognise that some add-ons may rely on incorrect bugs or behaviour.
```

However, preserving a bug to maintain backwards compatibility significantly hampers NVDA's development.
If a public function was documented to do X but was actually doing Y, changing it to correctly do X is considered a bug fix, not a breaking change.

> * Conversely, if a function that was documented to do X, and was correctly doing X, is changed to instead do Y - this is considered a breaking change.

Reviewer: Perhaps this shouldn't be its own list item. The "Conversely" refers to (negates) an example condition given in a sub-part of the prior list item.

Also suggesting a minor punctuation change, though I'm not sure if a colon or comma would be better there.

```suggestion
Conversely, if a function that was documented to do X, and was correctly doing X, is changed to instead do Y: this is considered a breaking change.
```

Author: Agreed it is a bit redundant. Removed it.

##### Security improvements

Security takes precedence over backward compatibility.

> We will make every effort to document these changes clearly in the changelog.

Author: ```suggestion
We will make every effort to document these changes clearly (see below).
```

* **Examples:** Privilege escalation or insecure file handling.

> * **Policy:** These are permitted in any release but are primarily deployed in patch releases (e.g. 2026.1.2).

Author: ```suggestion
* **Policy:** These are permitted in any release but are primarily deployed in patch releases (e.g. 2026.1.2).
* If an urgent and important security fix requires an API signature change in a minor or patch release, we will not treat the release as an 'API-breaking release' (i.e. we will not force a global add-on compatibility update). 
Rather than disabling the entire ecosystem, we will limit the impact to a runtime incompatibility in those add-ons consuming the insecure API.
```

#### API deprecation strategy

We aim to avoid silent breaks of add-on code.

> When removing or changing APIs in a breaking release (e.g. 2026.1):

Author: ```suggestion
When removing or changing APIs in a breaking release (e.g. 2026.1) or due to an urgent security improvement (e.g. 2026.1.2):
```
